### ISSUE TITLE
Server side searchable select clears value after selecting

### ISSUE THREAD

**Issue by lucasromanojf** [livewire/flux#581]
Hello,

I'm trying to create a searchable select where options are filtered server side. 

The problem is that after searching and selecting an option, the selected value is immediately cleared. 

I guess this happens because the search box is cleared after selecting an option, which makes the options being cleared and also the selected value (a cascade effect).

What is the correct way to achieve what I am looking for?

Here is an example Volt component where the issue can be reproduced:

```
<?php

use function Livewire\Volt\{computed, state};

state([
    'search' => '',
    'industry' => '',
    'industries' => [
        ['value' => 1, 'label' => 'Photography'],
        ['value' => 2, 'label' => 'Design services'],
        ['value' => 3, 'label' => 'Web development'],
        ['value' => 4, 'label' => 'Accounting'],
        ['value' => 5, 'label' => 'Legal services'],
        ['value' => 6, 'label' => 'Consulting'],
        ['value' => 7, 'label' => 'Other'],
    ],
]);

$options = computed(function () {
    if (! filled($this->search)) {
        return collect();
    }

    return collect($this->industries)
        ->filter(fn ($option) => str_contains($option['label'], $this->search))
        ->values();
});
?>

<div>
    <flux:select
        wire:model="industry"
        variant="listbox"
        searchable
        :filter="false"
        placeholder="Select an industry"
    >
        <x-slot name="search">
            <flux:select.search wire:ignore wire:model.live.debounce.500ms="search" placeholder="Search..." />
        </x-slot>

        @foreach ($this->options as $option)
            <flux:option wire:key="industry-{{ $option['value'] }}" value="{{ $option['value'] }}">{{ $option['label'] }}</flux:option>
        @endforeach
    </flux:select>
</div>
```

**Comment by jeffchown** [comment]
@lucasromanojf Your code clears all of the options when the `search` is empty, and per your comment "because the search box is cleared after selecting an option", this will cause the list of valid options to be empty right after a user has selected one.

It looks like changing your computed $options to this:

```
$options = computed(function () {
    // if (! filled($this->search)) {
    //     return collect();
    // }

    return collect($this->industries)
        ->filter(fn ($option) => str_contains($option['label'], $this->search))
        ->values();
});
```
fixes the issue.

**Comment by lucasromanojf** [comment]
> @lucasromanojf Your code clears all of the options when the `search` is empty, and per your comment "because the search box is cleared after selecting an option", this will cause the list of valid options to be empty right after a user has selected one making their choice effectively invalid.
> 
> It looks like changing your computed $options to this:
> 
> ```
> $options = computed(function () {
>     // if (! filled($this->search)) {
>     //     return collect();
>     // }
> 
>     return collect($this->industries)
>         ->filter(fn ($option) => str_contains($option['label'], $this->search))
>         ->values();
> });
> ```
> fixes the issue.

Hello,

The problem is that by doing this I will always need to load all options when the page loads. If I have many options that will be a problem.

**Comment by lucasromanojf** [comment]
I was able to fix it by doing this way:

```
<?php

use function Livewire\Volt\{computed, state};

state([
    'search' => '',
    'industry' => '',
    'industries' => [
        ['value' => 1, 'label' => 'Photography'],
        ['value' => 2, 'label' => 'Design services'],
        ['value' => 3, 'label' => 'Web development'],
        ['value' => 4, 'label' => 'Accounting'],
        ['value' => 5, 'label' => 'Legal services'],
        ['value' => 6, 'label' => 'Consulting'],
        ['value' => 7, 'label' => 'Other'],
    ],
]);

$options = computed(function () {
    if (filled($this->industry)) {
        return collect($this->industries)
            ->filter(fn ($option) => $option['value'] == $this->industry || str($option['label'])->contains($this->search, true))
            ->values();
    }

    if (! filled($this->search)) {
        return collect();
    }

    return collect($this->industries)
        ->filter(fn ($option) => str($option['label'])->contains($this->search, true))
        ->values();
});
?>

<div>
    <flux:select
        wire:model="industry"
        variant="listbox"
        searchable
        :filter="false"
        placeholder="Select an industry"
    >
        <x-slot name="search">
            <flux:select.search wire:ignore wire:model.live.debounce.500ms="search" placeholder="Search..." />
        </x-slot>

        @foreach ($this->options as $option)
            <flux:option wire:key="industry-{{ $option['value'] }}" value="{{ $option['value'] }}">{{ $option['label'] }}</flux:option>
        @endforeach
    </flux:select>
</div>
```

**Comment by jeffchown** [comment]
@lucasromanojf You can also set the $search == the chosen 'label' via an `updated` hook like this:

```
updated([
    'industry'
        => fn ($value)
            => $this->search =
                    collect($this->industries)->firstWhere("value", $value)["label"]
]);
```

**Comment by jeffchown** [comment]
@lucasromanojf Quick note, I would use `strtolower()` in your comparison code in your computed method - right now your search is case-sensitive.

**Comment by lucasromanojf** [comment]
> @lucasromanojf Quick note, I would use `strtolower()` in your comparison code in your computed method - right now your search is case-sensitive.

No problem, it is not a real code, just wanted to illustrate the main issue. A real use case would probably use eloquent with mysql which would be case insensitive by default

**Comment by calebporzio** [comment]
Thanks for the report. This behavior was actually intended. I tested with native <select> elements during development and observed that if you remove an option element that is selected, it clears the selection.

However, I see that this behavior prohibits using server-side driven/changing options list without wiping the selected value.

I added a little cache to preserve the removed option elements in memory if they aren't in the DOM (with some garbage collection code of course).

Thanks for the report.

This update will be in the next release. thanks.
